src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.cpp

```diff
@@ -227,6 +226,84 @@ tnsr::A<DataType, SpatialDim, Frame> compute_spacetime_normal_vector(
   return spacetime_normal_vector;
 }
 
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::ii<DataType, SpatialDim, Frame> compute_extrinsic_curvature(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::iJ<DataType, SpatialDim, Frame>& deriv_shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
+    const tnsr::ijj<DataType, SpatialDim, Frame>&
+        deriv_spatial_metric) noexcept {
+  const DataType half_over_lapse = 0.5 / get<>(lapse);
+
+  auto extrinsic_curvature =
+      make_with_value<tnsr::ii<DataType, SpatialDim, Frame>>(lapse, 0.0);
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = i; j < SpatialDim; ++j) {  // Symmetry
+      for (size_t k = 0; k < SpatialDim; ++k) {
+        extrinsic_curvature.get(i, j) +=
+            shift.get(k) * deriv_spatial_metric.get(k, i, j) +
+            spatial_metric.get(k, i) * deriv_shift.get(j, k) +
+            spatial_metric.get(k, j) * deriv_shift.get(i, k);
+      }
+      extrinsic_curvature.get(i, j) -= dt_spatial_metric.get(i, j);
+      extrinsic_curvature.get(i, j) *= half_over_lapse;
+    }
+  }
+
+  return extrinsic_curvature;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> compute_gauge_source_function(
+    const Scalar<DataType>& lapse, const Scalar<DataType>& dt_lapse,
+    const tnsr::i<DataType, SpatialDim, Frame>& deriv_lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::I<DataType, SpatialDim, Frame>& dt_shift,
+    const tnsr::iJ<DataType, SpatialDim, Frame>& deriv_shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const Scalar<DataType>& trace_extrinsic_curvature,
+    const tnsr::i<DataType, SpatialDim, Frame>&
+        trace_christoffel_last_indices) noexcept {
+  DataType one_over_lapse = 1.0 / get<>(lapse);
+  auto gauge_source_function =
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(lapse, 0.0);
+
+  // Temporary to avoid more nested loops.
+  auto temp0 = dt_shift;
```

Comment:
why not just `temp`?

Reply:
Oops, there used to be a temp1 that I eliminated...